### ISSUE TITLE
CFP: BGP Unnumbered support for BGP CPlane

### ISSUE THREAD

**Issue by YutaroHayakawa** [cilium/cilium#22132]
## Cilium Feature Proposal

**Is your feature request related to a problem?**

Currently, users need to specify BGP peers with IP addresses. Thus, it needs explicit knowledge about the peer configuration. This is automation unfriendly since users need to import such knowledge from an external system like DC-wide IPAM.

**Describe the feature you'd like**

In modern DC networks, the technique called "BGP Unnumbered" is often used, which typically allows users to specify peers with network interface names like `eth0`. This is more automation-friendly than specifying a peer with an IP address because users don't need to know the peer's IP address and only know which interface it is connected to.

BGP Unnumbered consists of multiple techniques (there are several ways, but below I'll describe it with FRR's implementation)
1. BGP peering with IPv6 link-local address (users don't need to assign IP address for peering, this is why it is called "unnumbered").
2. Exchanging IPv4 prefix with IPv6 next hop ([RFC8960](https://datatracker.ietf.org/doc/rfc8950/))
3. Automatic neighbor discovery with IPv6 RA (users don't need to know the peer IP address, we can find it automatically with the IPv6 discovery mechanism).
4. Installing IPv4 prefix with IPv6 next hop to DPlane's FIB

**(Optional) Describe your proposed solution**

The configuration knob of `CiliumBGPPeeringPolicy` would look like the below. It may be worth having an annotation to override it per-node for the case nodes don't have a consistent device naming scheme.

```
apiVersion: "cilium.io/v2alpha1"
kind: CiliumBGPPeeringPolicy
metadata:
 name: 01-bgp-peering-policy
spec: # CiliumBGPPeeringPolicySpec
 nodeSelector:
   matchLabels:
     bgp-policy: a
 virtualRouters: # []CiliumBGPVirtualRouter
 - localASN: 64512
   exportPodCIDR: true
   neighbors: # []CiliumBGPNeighbor
    - peerInterface: 'eth0' <= new
      # peerAddress: 'fc00:f853:ccd:e793::50/128'
      peerASN: 64512
```

- For 1 and 2, GoBGP already supports it (including the configuration knob to specify peer with interface name). We just need to wire up it with Cilium's configuration knob. The only concern is in BGP unnumbered, users typically assign the node's IP address to the loopback address, keep physical devices unnumbered, and advertise the loopback interface's address with BGP. I don't know whether it works with the features such as KPR, which needs explicit IP address assignment to devices.
- For 3, GoBGP doesn't support RA. Thus, we need to support that with Cilium.
- ~~For 4, from v5.3, Linux kernel supports it with [nexthop object](https://lwn.net/Articles/763950/), but for older kernels we need another way. There is a well-known workaround for this. "Forwarding with RFC 5549" section of [this article](https://www.oreilly.com/library/view/bgp-in-the/9781491983416/ch04.html) describes it well.~~ Currently, Cilium doesn't support importing the route. Thus we don't have to worry about this for now.

**Misc**
Also suggested by community user: https://github.com/cilium/cilium/issues/20653#issuecomment-1304448173

**Comment by matjazp** [comment]
Any news regarding BGP unnumbered support in Cilium?

**Comment by knfoo** [comment]
I will repeat the above question - is this going to be a thing ?

**Comment by YutaroHayakawa** [comment]
For now, myself don't have any plan to work on this, but as always, contributions are welcome.

**Comment by maartenvdezz** [comment]
This would be very handy indeed. Now I use FRR on the node as a router in the middle (because FRR can do the unnumbered ipv6 peering)

**Comment by Nenq** [comment]
very important feature for on-premise deployments.

**Comment by maartenvdezz** [comment]
Unnumbered peering (using IPv6 link local), it's a standard for datacenters, would be awesome!

**Comment by modzilla99** [comment]
> This would be very handy indeed. Now I use FRR on the node as a router in the middle (because FRR can do the unnumbered ipv6 peering)

How do you do that? How can Cilium peer with FRR on the same node. FRR does not allow to peer with the node itself.

**Comment by knfoo** [comment]
I have this running in production clusters. 

<img width="161" height="281" alt="Image" src="https://github.com/user-attachments/assets/d55f9922-6d3a-4eaf-a306-6b328e0ca9eb" />

Frr needs to look something like this:

```
router bgp {{NODE_ASN}}
  bgp router-id {{NODE_IP}}
  bgp bestpath as-path multipath-relax
  bgp allow-martian-nexthop
  no bgp ebgp-requires-policy
  neighbor LEAFS peer-group
  neighbor LEAFS remote-as external
  neighbor LEAFS timers 3 9
{%- for n in ports %}
  neighbor {{n}} interface peer-group LEAFS
{%- endfor %}
  neighbor CILIUM peer-group
  neighbor CILIUM remote-as external
  neighbor CILIUM capability extended-nexthop
  neighbor CILIUM graceful-restart
  neighbor CILIUM ebgp-multihop
  neighbor CILIUM update-source lo
  neighbor 127.0.0.1 peer-group CILIUM
  address-family ipv4 unicast
    redistribute kernel route-map denycidr
    redistribute connected route-map loopbacks
    neighbor CILIUM next-hop-self
    neighbor LEAFS allowas-in 1
    maximum-paths 64 
  exit-address-family
!
```

And cilium peering config:

``` 
apiVersion: "cilium.io/v2"
kind: CiliumBGPClusterConfig
metadata:
  name: rack0
spec:
  nodeSelector:
    matchLabels:
      rack: rack0
  bgpInstances:
  - localASN: 65013
    name: Cilium
    peers:
    - peerAddress: "127.0.0.1"
      name: localhost
      peerASN: NODE_ASN
``` 

And lastly on the k8s node you need to have a route on your `lo` device.

```
ip r a 127.0.0.1/32 dev lo
```

**Comment by maartenvdezz** [comment]
> I have this running in production clusters.

@knfoo use something similar but it shouldn't be necessary to run a dedicated frr instance with ebgp peering for this.

**Comment by knfoo** [comment]
@maartenvdezz true but does it not also depend on the use-case ?

Our nodes have no network before they have FRR running and advertising the node IP to BPG via ipv6 link local. 
And because of this there is no k8s and again no cilium, making a bootstrap process convoluted if we want to do this with cilium only. Not sure if cilium standalone even supports BGP peering 🤔

**Comment by modzilla99** [comment]
@knfoo we have exactly the same setup, thanks for sharing! It helped me a lot! Did not know about `bgp allow-martian-nexthop`...

**Comment by masap** [comment]
@YutaroHayakawa Hello, I am a first-time contributor to the `Cilium` project. I learned about this issue through @yassan and would like to take on the challenge of fixing it.

I am planning to fix this issue in two steps:

1. Add processing for receiving IPv6 RAs
2. Advertise IPv4 prefixes with an IPv6 next hop (BGP Control plane and Datapath)

First, I am thinking of submitting a PR for step 1. Does that sound good to you?

**Comment by YutaroHayakawa** [comment]
Hi @masap thanks for taking this! Since it's a sizable feature, it is better for us to discuss how to implement it. If you're still not familiar with BGP Control Plane code base, I encourage you to take a look at that first. (perhaps making small-size improvement to familiarize yourself).

Regarding 1. which RA implementation are you planning to use? If you don't have strong opinion, I'd recommend [go-ra](https://github.com/YutaroHayakawa/go-ra).

For 2, BGP Control Plane already advertises extended-nexthop capability (IPv4 NLRI, IPv6 nexthop). We don't need any datapath change as we only advertise route.

**Comment by masap** [comment]
Hello @YutaroHayakawa, thank you for your kind reply!

> Since it's a sizable feature, it is better for us to discuss how to implement it. If you're still not familiar with BGP Control Plane code base, I encourage you to take a look at that first. (perhaps making small-size improvement to familiarize yourself).

Yes, I am currently studying BGP (GoBGP) and eBPF with the assistance of Claude Code.

> Regarding 1. which RA implementation are you planning to use? If you don't have strong opinion, I'd recommend go-ra.

It seems that a feature to register the neighbor's Global and Link-local IPv6 addresses into the Cilium `statedb` upon receiving an IPv6 RA already exists, so I am planning to use it. The content of the IPv6 RA packet contains the Global address, but since the source is a link-local address, I believe that can be utilized. My reasoning is based on RFC 4861 Section 6.1.2, Validation of Router Advertisement Messages [1], which states, "IP Source Address is a link-local address." I have verified this through the following steps, but is my understanding correct?

**Boot environment**
```
$ make kind-bgp-service
$ make kind-bgp-service-apply-bgp
```

**Check IPv6 addresses of router0 NIC**
```
$ docker exec -it clab-bgp-cplane-dev-service-router0 ip addr show dev net0
36: net0@if35: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9500 qdisc noqueue state UP group default 
    link/ether aa:c1:ab:23:5a:47 brd ff:ff:ff:ff:ff:ff link-netnsid 2
    altname clab-o-597cd933d54d93ee
    inet 10.0.1.1/24 scope global net0
       valid_lft forever preferred_lft forever
    inet6 fd00:10:0:1::1/64 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::a8c1:abff:fe23:5a47/64 scope link 
       valid_lft forever preferred_lft forever
```

**Enable RA**
```
$ docker exec -it clab-bgp-cplane-dev-service-router0 vtysh \
  -c 'configure terminal' \
  -c 'interface net0' \
  -c '  ipv6 nd ra-interval 10' \
  -c '  no ipv6 nd suppress-ra' \
  -c '!'
```

**Confirm RA reception on Linux**
```
$ docker exec -it bgp-cplane-dev-service-control-plane ip neigh show dev net0
10.0.1.1 lladdr aa:c1:ab:23:5a:47 REACHABLE
fd00:10:0:1::1 lladdr aa:c1:ab:23:5a:47 router REACHABLE # Global IPv6 address of router0
fe80::a8c1:abff:fe23:5a47 lladdr aa:c1:ab:23:5a:47 router STALE # Link local IPv6 address of router0
```

**Confirm RA reception on Cilium statedb**
```
$ CILIUM_POD=$(kubectl --context kind-bgp-cplane-dev-service get pods -n kube-system \
  -l k8s-app=cilium -o wide | grep "control-plane" | awk '{print $1}')
$ echo $CILIUM_POD
cilium-82wkh
$ kubectl --context kind-bgp-cplane-dev-service exec -n kube-system $CILIUM_POD -- cilium-dbg statedb | jq -cr '.neighbors[]'
{"LinkIndex":2,"IPAddr":"172.18.0.1","HardwareAddr":"16:d3:f7:a3:cf:c3","Type":1,"State":2,"Flags":0,"FlagsExt":0}
{"LinkIndex":2,"IPAddr":"172.18.0.2","HardwareAddr":"7e:c5:6d:ec:37:af","Type":1,"State":2,"Flags":0,"FlagsExt":0}
{"LinkIndex":35,"IPAddr":"10.0.1.1","HardwareAddr":"aa:c1:ab:23:5a:47","Type":1,"State":2,"Flags":0,"FlagsExt":0}
{"LinkIndex":35,"IPAddr":"fd00:10:0:1::1","HardwareAddr":"aa:c1:ab:23:5a:47","Type":1,"State":2,"Flags":128,"FlagsExt":0} # Global IPv6 address of router0
{"LinkIndex":35,"IPAddr":"fe80::a8c1:abff:fe23:5a47","HardwareAddr":"aa:c1:ab:23:5a:47","Type":1,"State":4,"Flags":128,"FlagsExt":0} # Link local IPv6 address of router0
```

> For 2, BGP Control Plane already advertises extended-nexthop capability (IPv4 NLRI, IPv6 nexthop). We don't need any datapath change as we only advertise route.

Thank you for the information. In that case, I think I will just work on step 1.

[1] https://www.rfc-editor.org/rfc/rfc4861.txt

**Comment by YutaroHayakawa** [comment]
> It seems that a feature to register the neighbor's Global and Link-local IPv6 addresses into the Cilium statedb upon receiving an IPv6 RA already exists, so I am planning to use it.

That implementation is good for Cilium => Peer direction, but one caveat here is without advertising RA from Cilium, the peer cannot discover Cilium's link-local address, so the peer cannot connect to Cilium proactively. I'd say it's ok to start without RA advertisement. In that case, we should document this caveats.

> The content of the IPv6 RA packet contains the Global address, but since the source is a link-local address, I believe that can be utilized. My reasoning is based on RFC 4861 Section 6.1.2, Validation of Router Advertisement Messages [1], which states, "IP Source Address is a link-local address." I have verified this through the following steps, but is my understanding correct?

I think you're right.

> Yes, I am currently studying BGP (GoBGP) and eBPF with the assistance of Claude Code.

FYI, if you're considering using GenAI for your contribution, [here](https://github.com/cilium/community/blob/main/AI-POLICY.md)'s the policy that you should know.

**Comment by masap** [comment]
> That implementation is good for Cilium => Peer direction, but one caveat here is without advertising RA from Cilium, the peer cannot discover Cilium's link-local address, so the peer cannot connect to Cilium proactively. I'd say it's ok to start without RA advertisement. In that case, we should document this caveats.

Ah yes, as you pointed out, it is a one-way implementation. While it will eventually need to be bidirectional, I plan to proceed with this implementation for now.

> FYI, if you're considering using GenAI for your contribution, here's the policy that you should know.

Sure, I'll take a look.

**Comment by ecbaldwin** [comment]
Hi @YutaroHayakawa @masap — I just revisited this issue to reread it and saw
the recent activity. Coincidentally, I also recently began working on the
link-local peering piece (CFP item 1). Wanted to surface it before opening a PR
in case it's worth collaborating.

Rough shape of what I started on a [branch off main](https://github.com/ecbaldwin/cilium/tree/link-local-peering):

- A new peerInterface field (using the name from your CFP example) on the v2
  BGP CRDs.
- Both flavors wired up: explicit fe80::X paired with an interface, and the
  unnumbered case where only the interface is given. The unnumbered side leans
  on gobgp's existing NeighborInterface and its built-in IPv6 ND — basically
  wiring up what the CFP's "GoBGP already supports it" note pointed to.
- CEL validation, per-node override, tests, regenerated CRDs.

I have only begun to setup this code in our labs. We have cilium hosts peering
with DPUs and I have a little bit of internal work to do to get this off of the
ground, hence the lack of a mention here and/or PR created.

It doesn't touch items 2 or 3 yet. Those would be complimentary to this.

A few things I'd love feedback on:

1. PR item 1 standalone now, or hold until the RA work lands so they go
   together?
2. One commit covering both modes, or split?
3. Anything in the v2 API shape worth flagging up front? The CFP example was on
   v2alpha1.

**Comment by YutaroHayakawa** [comment]
Thanks for the heads up @ecbaldwin. Seems like both of you and @masap are working on the same thing at the same time. What I'd suggest now is coordinate among you and unify your work (e.g. you guys can collaborate on the same branch) make the final outcome usable for both of your use cases. I don't want to make this like you both have PRs and we need to choose one of them.

**Comment by YutaroHayakawa** [comment]
> A new peerInterface field (using the name from your CFP example) on the v2
BGP CRDs.

Since I write this issue, the `autoDiscovery` stanza was introduced. This covers any BGP peer auto discovery mechanism, so instead of introducing a new [peeringInterface](https://github.com/ecbaldwin/cilium/blob/166bc85b50e1b9749e616908f79c2ce33244f415/pkg/k8s/apis/cilium.io/v2/bgp_cluster_types.go#L151-L162).

```
autoDiscovery:
  mode: Unnumbered
  unnumbered:
    interface: "net0"
```

This might be useful for you as well @masap.

**Comment by masap** [comment]
Ah, it looks like @ecbaldwin has already submitted a PR. It takes the approach of reconciling the neighbor table, and I think it is a good PR. I believe @ecbaldwin can produce better results working alone than if I were to intervene, so I will leave this issue and look for another one. Good luck!

**Comment by ecbaldwin** [comment]
I got around to this in the last week and refined by earlier branch with a more concrete proposal. I pivoted it to match @YutaroHayakawa 's proposal for the shape of the autoDiscovery.

gobgp does most of the work, but note that I had to make a bug fix to the vendored code (both v3 and v4) to make it worked through the grpc code path. I will follow up with a link to the gobgp here soon.

I also added a commit to the PR to send RAs from cilium on the link. gobgp does not do this and FRR, at least, would not accept an unnumbered peer from which it did not receive an RA. Cilium did not need any network permissions that it did not already have.

**Comment by coelho** [comment]
Chiming in on this since I'm really interested in this Issue/PR, but I'm confused about one aspect:

> [@maartenvdezz](https://github.com/maartenvdezz) true but does it not also depend on the use-case ?
> 
> Our nodes have no network before they have FRR running and advertising the node IP to BPG via ipv6 link local. And because of this there is no k8s and again no cilium, making a bootstrap process convoluted if we want to do this with cilium only. Not sure if cilium standalone even supports BGP peering 🤔

1. It seems there would be a chicken egg situation when trying to use this feature for a fully unnumbered Cilium deployment.
2. The obvious workaround is to use fallback routes to each leaf that allow Kubelet and Cilium to bootstrap.
3. However, if you’re relying on BGP for fault tolerance (likely, considering that's the point of this feature), fallback routes lead to a lack of redundancy during bootstrap.
4. At that point, we're back to FRR.

Are there any ideas for addressing this? It would be great to have a documented best practice for this feature.

**Comment by maartenvdezz** [comment]
@coelho yes I would add something like a separate helper for Cilium BGP (which can be run as quadlet container) so that it can advertise a dummy node IP before Kubernetes and Cilium is bootstrapped.

I don't know how this is implemented in Isovalent Enterprise, but I do know that unnumbered peering is included there and I assume they have thought about this.

**Comment by coelho** [comment]
> [@coelho](https://github.com/coelho) yes I would add something like a separate helper for Cilium BGP (which can be run as quadlet container) so that it can advertise a dummy node IP before Kubernetes and Cilium is bootstrapped.

I agree. Cilium BGP also needs to survive rollouts and CrashLoopBackOff (for registry pulls and disaster scenarios) to prevent nodes from losing connectivity. I believe that requires the BGP daemon to run outside of Kubernetes.

Without a solution, the only alternative is to run FRR on the host outside of Cilium, which makes this feature effectively useless.

> I don't know how this is implemented in Isovalent Enterprise, but I do know that unnumbered peering is included there and I assume they have thought about this.

definitely curious!
